### PR TITLE
Add WebClient based sender for Zipkin

### PR DESCRIPTION
**Pull request by StefanBratanov**
## PR Description

- Introduce abstract `HttpSender` class
- Introduce `ZipkinWebClientSender`
- Add `ZipkinWebClientSender` bean to configuration

related to #30661

### CODE REVIEW

**Comment by wilkinsona** [comment]
Nothing at the moment, @StefanBratanov. Moritz is out for a few weeks on paternity leave so it may take us a little while to get to it. 3.0 doesn't GA until November so there's no rush at this stage.

**Comment by StefanBratanov** [comment]
> Nothing at the moment, @StefanBratanov. Moritz is out for a few weeks on paternity leave so it may take us a little while to get to it. 3.0 doesn't GA until November so there's no rush at this stage.

Sure. No worries. Thank you for quick update. Congratulations to Moritz. 🙂

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinConfigurations.java [hunk 1]
```diff
@@ -56,6 +62,20 @@ Sender urlConnectionSender(ZipkinProperties properties) {
...
+		@ConditionalOnMissingBean
+		@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.REACTIVE)
+		@ConditionalOnBean(WebClient.Builder.class)
+		@ConditionalOnMissingClass("zipkin2.reporter.urlconnection.URLConnectionSender")
+		Sender webClientSender(ZipkinProperties properties, WebClient.Builder webClientBuilder) {
+			HttpClient client = HttpClient.create()
+					.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, (int) properties.getConnectTimeout().toMillis())
+					.doOnConnected((conn) -> conn
+							.addHandlerLast(new ReadTimeoutHandler((int) properties.getReadTimeout().toSeconds())));
+			WebClient webClient = webClientBuilder.clientConnector(new ReactorClientHttpConnector(client)).build();
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinConfigurations.java:75]: This introduces a hard dependency on Netty, "just" for the sake of setting a timeout.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/HttpSender.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	public Call<Void> sendSpans(List<byte[]> encodedSpans) {
+		if (this.closed) {
+			throw new ClosedSenderException();
+		}
+		return sendSpans(BytesMessageEncoder.JSON.encode(encodedSpans));
+	}
+
+	abstract static class HttpPostCall extends Call.Base<Void> {
+
+		byte[] body;
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/HttpSender.java:102]: Can this be private final?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/HttpSender.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ * compression with gzip.
+ *
+ * @author Moritz Halbritter
+ * @author Stefan Bratanov
+ */
+abstract class HttpSender extends Sender {
+
+	private static final DataSize MESSAGE_MAX_BYTES = DataSize.ofKilobytes(512);
+
+	volatile boolean closed;
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/HttpSender.java:46]: Can this be private?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSender.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			super(body);
+			this.endpoint = endpoint;
+			this.webClient = webClient;
+		}
+
+		@Override
+		void post(byte[] body, HttpHeaders defaultHeaders) {
+			this.webClient.post().uri(this.endpoint).headers((headers) -> headers.addAll(defaultHeaders))
+					.bodyValue(body).retrieve().toBodilessEntity().subscribe((__) -> {
+					}, (__) -> {
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSender.java:61]: This looks wrong, the errors will be suppressed. The webclient and the Zipkin Sender interface both support callbacks, maybe we could refactor this to fix the error handling?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSenderTests.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+	public static MockWebServer mockBackEnd;
+
+	private static String ZIPKIN_URL;
+
+	private ZipkinWebClientSender sut;
+
+	@BeforeAll
+	static void beforeAll() throws IOException {
+		mockBackEnd = new MockWebServer();
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSenderTests.java:56]: Is this integration testing the only way to test webclients? RestTemplate has a built-in test infrastructure, I wonder if WebClient has the same?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinConfigurations.java [hunk 6]
```diff
@@ -56,6 +62,20 @@ Sender urlConnectionSender(ZipkinProperties properties) {
 					.build();
 		}
 
+		@Bean
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinConfigurations.java:65]: I have fixed a bug in `main` regarding the auto-configuration, I guess you'll have to rebase.

**Review by mhalbritter — CHANGES_REQUESTED** [review]

**Comment by mhalbritter** [comment]
Hey, thanks for the contribution. I'm back and have some comments on the code.

**Comment by StefanBratanov** [comment]
> Hey, thanks for the contribution. I'm back and have some comments on the code.

Welcome back! Thank you. I will have a look.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinRestTemplateSender.java [hunk 7]
```diff
@@ -16,153 +16,71 @@
...
-		}
-
-		private byte[] compress(byte[] input) throws IOException {
-			ByteArrayOutputStream result = new ByteArrayOutputStream();
-			try (GZIPOutputStream gzip = new GZIPOutputStream(result)) {
-				gzip.write(input);
-			}
-			return result.toByteArray();
+		protected void doEnqueue(Callback<Void> callback) {
+			CompletableFuture.supplyAsync(this::doExecute).whenComplete((__, throwable) -> {
```
Comment (mhalbritter) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinRestTemplateSender.java:76]: We don't need the `CompletableFuture` in this case. The UrlConnectionSender implements the `enqueue` call without additional threading, so we should be fine here without introducing multi-threading.

**Comment by mhalbritter** [comment]
Thanks @StefanBratanov for the changes, the PR looks quite nice now!

> Hi Moritz, sorry for the delay. I did make a refactor based on the comments, which i agree with. I wasn't sure about handling the `WebClient` async call, but now i think it makes sense and is cleaner. I also changed the `ZipkinRestTemplateSender` async implementation to use a `CompletableFuture` since it was blocking before. I also added tests for both sync and async in both implementations.

I don't think we need that. The `UrlConnectionSender` from zipkin implements the enqueue call without additional threading, so we should be fine here without introducing multi-threading.

> I am not sure about initializing the `WebClient` in `ZipkinConfigurations`. I was using the `ReactorClientHttpConnector` which is based on` Netty`, so that i can honor the read and connect timeouts set in `ZipkinProperties`. What do you think is a better approach for initializing the `WebClient`? 

I have to double check if there's really no way to set timeouts without tying us directly on the implementation.

> Also i couldn't find a mock server similar to `MockRestServiceServer` for `WebClient`, so kept my old implementation. If there is one, I am happy to change it.

Yes, I was wrong, there's no such support in the Spring Framework. The recommendation is to use the [OkHttp MockServer](https://docs.spring.io/spring-framework/docs/6.0.0-SNAPSHOT/reference/html/web-reactive.html#webflux-client-testing), like you did.

**Comment by StefanBratanov** [comment]
Hi @mhalbritter thanks for review, I replaced the` CompletableFuture` implementation with a normal one. I will wait for you to check about the WebClient initialization. My logic was that `@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.REACTIVE)` in _ZipkinConfigurations_ if true, should mean that Netty should be on the classpath. I could be wrong though.

**Comment by mhalbritter** [comment]
Unfortunately, this is not the case. You can [use other servers for reactive applications](https://docs.spring.io/spring-boot/docs/3.0.0-SNAPSHOT/reference/htmlsingle/#web.reactive.reactive-server).

> Spring Boot includes support for the following embedded reactive web servers: Reactor Netty, Tomcat, Jetty, and Undertow

The best option would be to have timeout setter on the `WebClient.Builder`, but unfortunately this is not available [and won't be](https://github.com/spring-projects/spring-framework/issues/25115#issuecomment-677685984).

I would drop the timeout code in the configuration for now, and I will bring this up with the team if we can find a solution which works for all the connector implementations.

**Comment by StefanBratanov** [comment]
> Unfortunately, this is not the case. You can [use other servers for reactive applications](https://docs.spring.io/spring-boot/docs/3.0.0-SNAPSHOT/reference/htmlsingle/#web.reactive.reactive-server).
> 
> > Spring Boot includes support for the following embedded reactive web servers: Reactor Netty, Tomcat, Jetty, and Undertow
> 
> The best option would be to have timeout setter on the `WebClient.Builder`, but unfortunately this is not available [and won't be](https://github.com/spring-projects/spring-framework/issues/25115#issuecomment-677685984).
> 
> I would drop the timeout code in the configuration for now, and I will bring this up with the team if we can find a solution which works for all the connector implementations.

Got it. Makes sense. Good to know! The default builder will take care of initialising the appropriate connector. I removed the Netty connector.

**Comment by mhalbritter** [comment]
I've merged your PR. Thanks for your work and congratulations on your first contribution to Spring Boot!

**Pull request closed without merging**